\documentclass[slidestop]{beamer}

\title{$k$-mer profiling\\
  {\small Theory and Implementation}}
\providecommand{\myConference}{Work discussion}
\providecommand{\myDate}{Wednesday, 10 April 2013}
\author{Jeroen F. J. Laros}
\providecommand{\myGroup}{Leiden Genome Technology Center}
\providecommand{\myDepartment}{Department of Human Genetics}
\providecommand{\myCenter}{Center for Human and Clinical Genetics}
\providecommand{\lastCenterLogo}{
  \raisebox{-0.1cm}{
    \includegraphics[height = 1cm]{lgtc_logo}
  }
}
\providecommand{\lastRightLogo}{
  %\includegraphics[height = 0.7cm]{nbic_logo}
  %\includegraphics[height = 0.8cm]{nwo_logo_en}
  %\hspace{1.5cm}\includegraphics[height = 0.7cm]{gen2phen_logo}
}

\usetheme{lumc}

\begin{document}

% This disables the \pause command, handy in the editing phase.
%\renewcommand{\pause}{}

% Make the title page.
\bodytemplate

% First page of the presentation.
\section{Introduction}
\begin{frame}
  \frametitle{Raw sequencing data}

  When do we work with \emph{raw data}:
  \begin{itemize}
    \item Unknown reference.
    \item No time for analysis.
  \end{itemize}
  \bigskip
  \pause

  If the reference sequence is unknown, we can still do:
  \begin{itemize}
    \item Quality control.
    \item Coverage estimation.
  \end{itemize}
  \bigskip
  \pause

  Compare raw datasets:
  \begin{itemize}
    \item Quality control.
    \item Phylogeny.
    \item Metagenomics.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Counting $k$-mers}

  We choose a $k$ and count all occurrences of substrings of length $k$.
  \bigskip
  \pause

  The counts of these substrings serve as a fingerprint of the dataset.
  \bigskip
  \pause

  But, these counts contain a lot more information.
\end{frame}

\begin{frame}
  \frametitle{A $2$-mer profile}

  \begin{figure}[]
    \begin{center}
      \fbox{
        \newline

        \phantom{X}

        \newline
        % A>AC; C>T; T>AG; G>C
        \only<1>{\bt{
          \underline{AC}TAGACCACTTACTAGAGACTAGACCACCACTAGACCA\ldots}}\newline
        \only<2>{\bt{
          A\underline{CT}AGACCACTTACTAGAGACTAGACCACCACTAGACCA\ldots}}\newline
        \only<3>{\bt{
          AC\underline{TA}GACCACTTACTAGAGACTAGACCACCACTAGACCA\ldots}}\newline
        \only<4>{\bt{
          ACT\underline{AG}ACCACTTACTAGAGACTAGACCACCACTAGACCA\ldots}}\newline
        \only<5>{\bt{
          ACTA\underline{GA}CCACTTACTAGAGACTAGACCACCACTAGACCA\ldots}}\newline
        \only<6>{\bt{
          ACTAG\underline{AC}CACTTACTAGAGACTAGACCACCACTAGACCA\ldots}}\newline

        \phantom{X}

        \newline
      }
    \end{center}
    \caption{Indexing.}
    \label{}
  \end{figure}

  We use a sliding window to find all $k$-mers ($k = 2$).

  \begin{table}[]
    \begin{center}
      \begin{tabular}{ll|ll|ll|ll}
        AA & $0$ & CA & $0$ & GA &
          \only<-4>{$0$}\only<5->{\color{yellow}$1$\color{white}} & TA &
          \only<-2>{$0$}\only<3->{\color{yellow}$1$\color{white}}\\
        AC & \color{yellow}\only<-5>{$1$}\only<6->{$2$}\color{white} & CC &
          $0$ & GC & $0$ & TC & $0$\\
        AG & \only<-3>{$0$}\only<4->{\color{yellow}$1$\color{white}} & CG &
          $0$ & GG & $0$ & TG & $0$\\
        AT & $0$ & CT &
          \only<1>{$0$}\only<2->{\color{yellow}$1$\color{white}} & GT & $0$ &
          TT & $0$\\
      \end{tabular}
    \end{center}
    \caption{$2$-mer profile.}
    \label{}
  \end{table}

  Observations are stored in a table.
\end{frame}

\begin{frame}
  \frametitle{Counting $k$-mers}

  We choose a $k$ and count all occurrences of substrings of length $k$.
  \pause

  \begin{figure}
    \colorbox{white}{
      \includegraphics[width = \textwidth]{k-mer}
    }
    \vspace{-0.5cm}
    \caption{A profile of $3$-mer counts.} \label{fig:kmerprofile}
  \end{figure}

  In Figure~\ref{fig:kmerprofile} we see a part of $3$-mer counts; \bt{AAA}
  occurs $10$ times, \bt{AAC} occurs 20 times, etc.
\end{frame}

%\begin{frame}
%  \frametitle{Why $k$-mers?}
%
%  The usage of $k$-mers is appealing since:
%  \begin{itemize}
%    \item They are easy to work with.
%    \item The profiles are relatively small.
%    \item A lot of data is retained, despite the small size.
%    \item Can be used to make a database to match against.
%    \item Comparison between datasets takes little time compared to alignment.
%  \end{itemize}
%\end{frame}

\begin{frame}
  \frametitle{Choosing $k$}

  $k$ can not be too small:
  \begin{itemize}
    \item $k = 1$ will result in loss of all subsequence information.
    \item $k = 2$ will give you information about di-nucleotides.
    \item There is only one unique $10$-mer in Human (hg18).
  \end{itemize}
  \bigskip
  \pause

  But, $k$ can not be too large either:
  \begin{itemize}
    \item Almost all $18$-mers are unique in the Human genome.
    \item Since this is not error tolerant:
    \begin{itemize}
      \item Read errors.
      \item Assembly errors.
    \end{itemize}
  \end{itemize}
  \bigskip
  \pause

  We have a solution for this (explained later in this presentation).
\end{frame}

\section{$k$-mer profiles}
\begin{frame}
  \frametitle{Indexing}

  \begin{table}[]
    \begin{center}
      \begin{tabular}{c|c|c}
        nucleotide & binary  & decimal\\
        \hline
        \bt{A}     & \bt{00} & $0$\\
        \bt{C}     & \bt{01} & $1$\\
        \bt{G}     & \bt{10} & $2$\\
        \bt{T}     & \bt{11} & $3$\\
      \end{tabular}
    \end{center}
    \label{}
    \caption{Nucleotide encoding table.}
  \end{table}

  We use an additional trick to store profiles efficiently.
  \bigskip
  \pause

  First notice that we can encode a nucleotide in two bits.
\end{frame}

\begin{frame}
  \frametitle{Indexing}
  \begin{table}[]
    \begin{center}
      \begin{tabular}{c|c|r}
        substring & binary           & decimal\\
        \hline
        \bt{AAAA} & \bt{00 00 00 00} & $0$\\
        \bt{AAAC} & \bt{00 00 00 01} & $1$\\
        \vdots    & \vdots           & \vdots\\
        \bt{GATC} & \bt{10 00 11 01} & $141$\\
        \vdots    & \vdots           & \vdots\\
        \bt{TTTT} & \bt{11 11 11 11} & $255$\\
      \end{tabular}
    \end{center}
    \label{}
    \caption{Encoding strings.}
  \end{table}

  We can concatenate the \emph{binary encoding}.
  \bigskip
  \pause

  There is no need to store the substrings.
\end{frame}

\begin{frame}
  \frametitle{Comparing $k$-mer profiles}

  \vspace{-0.5cm}
  \begin{figure}
    \colorbox{white}{
      \includegraphics[width = \textwidth]{k-mer}
    }
    \smallskip

    \colorbox{white}{
      \includegraphics[width = \textwidth]{k-mer2}
    }
    \vspace{-0.5cm}
    \caption{Two profiles of $k$-mer counts.}
  \end{figure}
  \pause

  How to express this difference with one value.
\end{frame}

\begin{frame}
  \frametitle{Pairwise distance function}

  We use the following function:
  \begin{displaymath}
    f(x, y) = \frac{|x - y|}{(x + 1) (y + 1)}
  \end{displaymath}
  \pause

  Properties:
  \begin{itemize}
    \item $f(0, 1) = \frac12$
    \item $f(0, 1) > f(7, 8)$
  \end{itemize}
  \bigskip
  \bigskip
  \pause

  This is desirable:
  \begin{itemize}
    \item The fact that a $k$-mer is not present is more important than the
      number of times it is present.
    \item Differences in the low end of the spectrum are more important than
      ones at the high end.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Multiset distance function}

  Let $f$ be a function $f : \mathbb{R}_{\ge 0} \times \mathbb{R}_{\ge 0} 
                               \to \mathbb{R}_{\ge 0}$
  with finite supremum $M$ and the following properties:
  \begin{eqnarray*}
    f(x, y) &=& f(y, x) \hspace*{2cm}\mathrm{\ for\ all\ } x, y \in
      \mathbb{R}_{\ge 0}\\
    f(x, x) &=& 0 \hspace*{2.85cm}\mathrm{\ for\ all\ } x \in \mathbb{R}_{\ge
      0}\\
    f(x, 0) &\ge& {M}/2 \hspace*{2.3cm}\mathrm{\ for\ all\ } x \in
      \mathbb{R}_{> 0}\\
    f(x, y) &\le& f(x, z) + f(z, y) \hspace*{0.56cm}\mathrm{\ for\ all\ } x, y,
      z \in \mathbb{R}_{\ge 0}
  \end{eqnarray*}
  \smallskip
  \pause

  For a multiset $X$, let $S(X)$ denote its underlying set. For multisets $X,
  Y$ with $S(X),S(Y) \subseteq \{1, 2, \ldots, n\}$ we define 
  \smallskip

  \begin{displaymath}
    d_f(X, Y) = \frac{\sum_{i = 1}^n f(x_i, y_i)}{|S(X) \cup S(Y)| + 1}
  \end{displaymath}
  \bigskip
\end{frame}

\begin{frame}
  \frametitle{Strand balance}

  When analysing a dataset:
  \begin{itemize}
    \item We either see a $k$-mer or its reverse complement ($50\%$ chance of
      either).
    \item If sequenced in sufficient depth, we expect a balance between forward
      and reverse complement $k$-mers.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Strand balance}

  \begin{figure}[]
    \begin{center}
      \fbox{
        \begin{tabular}{l}
          \bt{ACTAGACCACTTAC\only<2->{\color{yellow}}TAGAGACTAGA\only<2->{\color{white}}CCACCACTAGACCA\ldots}\\
          \bt{TGATCTGGTGAATG\only<3->{\color{yellow}}ATCTCTGATCT\only<3->{\color{white}}GGTGGTGATCTGGT\ldots}\\
        \end{tabular}
      }
    \end{center}
    \caption{Double stranded DNA.}
    \label{}
  \end{figure}

  \onslide<4->{So, the number of times we see \bt{TAGAGACTAGA} must be roughly
  the same as for \bt{TCTAGTCTCTA}.}
\end{frame}

\begin{frame}
  \frametitle{Strand balance}

  How to calculate it:
  \begin{itemize}
    \item We can split a profile into a forward and a reverse complement
      profile.
    \item We can calculate the balance between these sub-profiles.
  \end{itemize}
  \bigskip
  \pause

  This is an estimation of ``sufficient coverage'':
  \begin{itemize}
    \item If the balance is good, the coverage is good.
    \begin{itemize}
      \item Does not work for strand specific protocols.
    \end{itemize}
  \end{itemize}
  \bigskip
  \pause

   We can balance the profile by adding $k$-mer counts to their reverse
   complement and vice versa.
\end{frame}


\begin{frame}
  \frametitle{Shrinking}

  \begin{figure}
    \fbox{
      \hspace{0.5cm}
      \begin{tabular}{cc}
        \bt{AAAA} & $1$ \\
        \bt{AAAC} & $7$ \\
        \bt{AAAG} & $0$ \\
        \bt{AAAT} & $9$ \\
        \bt{AACA} & $9$ \\
        \bt{AACC} & $8$ \\
        \bt{AACG} & $3$ \\
        \bt{AACT} & $7$ \\
        \vdots    & \vdots\\
      \end{tabular}
      \hspace{0.5cm}
      $\Rightarrow$
      \hspace{0.5cm}
      \begin{tabular}{ll}
        \\
        \bt{AAA} & $17$ \\
        \\
        \\
        \\
        \bt{AAC} & $27$ \\
        \\
        \\
        \vdots   & \vdots\\
      \end{tabular}
      \hspace{0.5cm}
    }
    \caption{Shrinking a profile.}
  \end{figure}
  \pause

  Works fine if the indexed sequences are large compared to $k$.
\end{frame}

\begin{frame}
  \frametitle{Smoothing}

  How do we compare multiple $k$-mer profiles?
  \medskip

  One sample might be sequenced deeper than the other, so the optimal $k$ might
  differ between comparisons.
  \pause

  \begin{figure}
    \fbox{
      \hspace{0.5cm}
      \begin{tabular}{ll}
        \bt{GTAAGTAA} & $0$ \\
        \bt{GTAAGTAC} & $1$ \\
        \bt{GTAAGTAG} & $0$ \\
        \bt{GTAAGTAT} & $1$ \\
        \bt{GTAAGTCA} & $9$ \\
        \bt{GTAAGTCC} & $8$ \\
        \bt{GTAAGTCG} & $3$ \\
        \bt{GTAAGTCT} & $7$ \\
      \end{tabular}
      \hspace{0.5cm}
      $\Rightarrow$
      \hspace{0.5cm}
      \begin{tabular}{ll}
        \\
        \bt{GTAAGTA}  & $2$ \\
        \\
        \\
        \bt{GTAAGTCA} & $9$ \\
        \bt{GTAAGTCC} & $8$ \\
        \bt{GTAAGTCG} & $3$ \\
        \bt{GTAAGTCT} & $7$ \\
      \end{tabular}
      \hspace{0.5cm}
    }
    \caption{Smoothing by collapsing sub-profiles.}
  \end{figure}
\end{frame}

\begin{frame}
  \frametitle{Smoothing}

  The function to determine when to smooth is a parameter:

  \begin{itemize}
    \item Median.
    \item Minimum.
    \item Average.
    \item \ldots
  \end{itemize}
  \bigskip

  This function has a threshold, which is also a parameter.
  \bigskip
  \pause

  We can index with a large $k$-mer size, this method automatically uses the
  optimal size when comparing.
\end{frame}

\section{Implementation: command line interface}
\begin{fframe}
  \frametitle{A programming library and command line interface}

  Uses a binary encoding of a $k$-mer as index:
  \begin{itemize}
    \item Only counts need to be stored.
  \end{itemize}
  \bigskip
  \pause

  Stored profiles can be:
  \begin{itemize}
    \item Loaded.
    \item Saved.
    \item Manipulated.
    \item Compared.
    \item \ldots
  \end{itemize}

  \vfill
  {\onslide<1->
    \bs{https://humgenprojects.lumc.nl/svn/k-mer/}
  }
\end{fframe}

\begin{frame}
  \frametitle{Indexing and querying}

  Generic functions:
  \begin{tabular}{@{\fakeitem}p{3cm}p{7cm}}
    index       & Make a profile from a FASTA file.\\
  \end{tabular}
  \bigskip
  \pause

  Summaries, querying, etc.:
  \begin{tabular}{@{\fakeitem}p{3cm}p{7cm}}
    info        & Generic info ($k$-mer length, total/non-zero counts).\\
    meanstd     & Mean and standard deviation of the count distribution.\\
    distr       & Histogram of the count distribution.\\
    showbalance & Show the balance of a profile.\\
    getcount    & Get the count for a specific $k$-mer.\\
  \end{tabular}
\end{frame}

\begin{frame}
  \frametitle{Profile manipulation}

  The following manipulaiton functions are implemented:
  \begin{tabular}{@{\fakeitem}p{3cm}p{7cm}}
    merge       & Merge two profiles.\\
    balance     & Balance a profile.\\
    positive    & Only keep counts that are positive in both profiles.\\
    scale       & Scale profiles such that the total number counts are equal.\\
    shrink      & Calculate a smaller profile.\\
    smooth      & Smooth two profiles by collapsing sub-profiles.\\
    shuffle     & Permute a profile.\\
  \end{tabular}
\end{frame}

\begin{frame}
  \frametitle{Comparison}

  For pairwise and multiple sample comparison:
  \begin{tabular}{@{\fakeitem}p{3cm}p{7cm}}
    diff        & Calculate the difference between two profiles.\\
    matrix      & Make a distance matrix any number of profiles.\\
  \end{tabular}
  \bigskip
  \pause

  The resulting \emph{distance matrix} can be visualised by a
  \emph{phylogenetic tree}, or by doing \emph{multidimensional scaling} (e.g.,
  PCA).
\end{frame}

\begin{frame}[fragile]
  \frametitle{Command line help}

  All commands are documented.
  \bigskip

  \begin{lstlisting}[numbers=none, basicstyle=\tiny, caption={Help for the
     diff option.}, language=none]
> python kMer.py diff -h
usage: kMer diff [-h] -i INPUT INPUT [-d] [-s SUMMARY] [-t THRESHOLD] [-b]
                 [-p] [-S] [-m] [-e] [-P PAIRWISE] [-n PRECISION]

Calculate the difference between two k-mer profiles.

optional arguments:
  -h, --help      show this help message and exit
  -i INPUT INPUT  pair of input files
  -d              scale down (default=False)
  -s SUMMARY      summary function for dynamic smoothing (int default=0)
  -t THRESHOLD    threshold for the summary function (int default=0)
  -b              balance the profiles (default=False)
  -p              use only positive values (default=False)
  -S              scale the profiles (default=False)
  -m              smooth the profiles (default=False)
  -e              use the euclidean distance metric (default=False)
  -P PAIRWISE     paiwise distance function for the multiset distance (int
                  default=0)
  -n PRECISION    number of decimals (int default=3)

  \end{lstlisting}
\end{frame}

\section{Programming library}
\begin{frame}
  \frametitle{Library for $k$-mer profiles}

  \bt{kLib}:
  \begin{itemize}
    \item Analyse a fasta file.
    \item Loading and saving of profiles.
    \item Merging.
    \item Balancing / splitting.
  \end{itemize}
  \bigskip
  \pause

  \bt{kDiffLib}:
  \begin{itemize}
    \item Smoothing.
    \item Make a comparison recipe.
  \end{itemize}
  \bigskip
  \pause

  \bt{kMer}:
  \begin{itemize}
    \item Command line interface.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{A more general library}

  These functions are not limited to $k$-mer profiles.
  \bigskip

  \bt{metrics}:
  \begin{itemize}
    \item Scaling.
    \item Extracting positive counts.
    \item Multiset distance function.
    \begin{itemize}
      \item Pairwise distance functions.
    \end{itemize}
    \item Euclidean distance function.
    \item Summary functions for the smoothing algorithm.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{API documentation}

  \begin{figure}[]
    \begin{center}
      \includegraphics[width=\textwidth]{kLib_api}
    \end{center}
    \caption{API documentation.}
    \label{}
  \end{figure}
\end{frame}

\section{Questions?}
\lastpagetemplate
\begin{fframe}
  \begin{center}
    Acknowledgements:
    \bigskip
    \bigskip

    Yahya Anvar

    Lusine Khachatryan

    Irina Pulyakhina

    Michiel van Galen

    Jaap van der Heijden

    Ken Kraaijeveld

    Walter Kosters

    Hendrik Jan Hoogeboom

    Johan den Dunnen
  \end{center}

  \vfill
  \bs{https://humgenprojects.lumc.nl/svn/k-mer/}
\end{fframe}

\end{document}
