\documentclass[slidestop]{beamer}

\title{$k$-mer programming library and toolkit}
\providecommand{\myConference}{Work discussion}
\providecommand{\myDate}{Wednesday, 8 August 2012}
\author{Jeroen F. J. Laros}
\providecommand{\myGroup}{Leiden Genome Technology Center}
\providecommand{\myDepartment}{Department of Human Genetics}
\providecommand{\myCenter}{Center for Human and Clinical Genetics}
\providecommand{\lastCenterLogo}{
  \raisebox{-0.1cm}{
    \includegraphics[height = 1cm]{lgtc_logo}
  }
}
\providecommand{\lastRightLogo}{
  %\includegraphics[height = 0.7cm]{nbic_logo}
  %\includegraphics[height = 0.8cm]{nwo_logo_en}
  %\hspace{1.5cm}\includegraphics[height = 0.7cm]{gen2phen_logo}
}

\usetheme{lumc}

\begin{document}

% This disables the \pause command, handy in the editing phase.
%\renewcommand{\pause}{}

% Make the title page.
\bodytemplate

% First page of the presentation.
\section{Introduction}
\begin{frame}
  \frametitle{Profiling}

  We frequently want to know something about raw datasets.
  \bigskip
  \pause

  Within a dataset (no or unknown reference):
  \begin{itemize}
    \item Quality control.
    \item Coverage estimation.
    \item Quality of a de novo assembly.
  \end{itemize}
  \bigskip
  \pause

  Between datasets:
  \begin{itemize}
    \item Quality control.
    \item Phylogeny.
    \item Metagenomics.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Counting $k$-mers}

  We choose a $k$ and count all occurrences of substrings of length $k$.
  \pause

  \begin{figure}
    \colorbox{white}{
      \includegraphics[width = \textwidth]{k-mer}
    }
    \vspace{-0.5cm}
    \caption{A profile of $3$-mer counts.} \label{fig:kmerprofile}
  \end{figure}

  In Figure~\ref{fig:kmerprofile} we see a part of $3$-mer counts; \bt{AAA}
  occurs $10$ times, \bt{AAC} occurs 20 times, etc.
\end{frame}

\begin{frame}
  \frametitle{Why $k$-mers?}

  The usage of $k$-mers is appealing since:
  \begin{itemize}
    \item They are easy to work with.
    \item Fingerprinting of samples is possible (even for a small $k$-mer
      length (around $10$)).
    \item Can be used to make a database to match against.
    \item Comparison between datasets takes little time compared to alignment.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Choosing $k$}

  $k$ can not be too small:
  \begin{itemize}
    \item $k = 1$ will result in loss of all subsequence information.
    \item $k = 2$ will give you information about di-nucleotides.
    \begin{itemize}
      \item But, pattern growth needs also position information.
    \end{itemize}
    \item There is only one unique $10$-mer in Human (hg18).
  \end{itemize}
  \bigskip
  \pause

  But, $k$ can not be too large either:
  \begin{itemize}
    \item Almost all $18$-mers are unique in the Human genome.
    \item Since this is not error tolerant:
    \begin{itemize}
      \item Read errors.
      \item Assembly errors.
    \end{itemize}
  \end{itemize}
  \bigskip
  \pause

  This is solved by collapsing profiles.
\end{frame}

\section{kMer}
\begin{fframe}
  \frametitle{A programming library and command line interface.}

  Uses a binary encoding of a $k$-mer as index:
  \begin{itemize}
    \item Only counts need to be stored.
  \end{itemize}
  \bigskip
  \pause

  Stored profiles can be:
  \begin{itemize}
    \item Loaded.
    \item Saved.
    \item Manipulated.
    \item Compared.
    \item \ldots
  \end{itemize}

  \vfill
  {\onslide<1->
    \bs{https://humgenprojects.lumc.nl/svn/k-mer/}
  }
\end{fframe}

\begin{frame}
  \frametitle{Command line interface.}

  \begin{tabular}{lp{7cm}}
    index       & Make a $k$-mer profile from a FASTA file.\\
    merge       & Merge two $k$-mer profiles.\\
    balance     & Balance a $k$-mer profile.\\
    showbalance & Show the balance of a $k$-mer profile.\\
    positive    & Only keep counts that are positive in both profiles.\\
    scale       & Scale profiles such that the total number of $k$-mers is 
      equal.\\
    smooth      & Smooth two profiles by collapsing sub-profiles.\\
    diff        & Calculate the difference between two $k$-mer profiles.\\
    matrix      & Make a distance matrix any number of $k$-mer profiles.\\
  \end{tabular}
\end{frame}

\begin{frame}[fragile]
  \frametitle{Command line help.}

  All commands are documented.
  \bigskip

  \begin{lstlisting}[numbers=none, basicstyle=\tiny, caption={Help for the
     diff option.}, language=none]
> python kMer.py diff -h
usage: kMer.py diff [-h] -i INPUT INPUT [-d] [-s SUMMARY] [-t THRESHOLD] [-b]
                    [-p] [-S] [-m] [-e] [-P PAIRWISE] [-n PRECISION]

Calculate the difference between two k-mer profiles.

optional arguments:
  -h, --help      show this help message and exit
  -i INPUT INPUT  pair of input files
  -d              scale down (default=False)
  -s SUMMARY      summary function for dynamic smoothing (int default=0)
  -t THRESHOLD    threshold for the summary function (int default=0)
  -b              balance the profiles (default=False)
  -p              use only positive values (default=False)
  -S              scale the profiles (default=False)
  -m              smooth the profiles (default=False)
  -e              use the euclidean distance metric (default=False)
  -P PAIRWISE     paiwise distance function for the multiset distance (int
                  default=0)
  -n PRECISION    number of decimals (int default=3)

  \end{lstlisting}
\end{frame}

\begin{frame}
  \frametitle{Balancing.}

  When analysing a dataset:
  \begin{itemize}
    \item We either see a $k$-mer or its reverse complement ($50\%$ chance of
      either).
    \item If sequenced in sufficient depth, we expect a balance between forward
      and reverse complement $k$-mers.
  \end{itemize}
  \bigskip
  \pause

  Funcionality in \bt{kLib}:
  \begin{itemize}
    \item We can split a profile into a forward and a reverse complement
      profile.
    \item We can calculate the balance between these sub-profiles.
    \item We can balance the profile by adding $k$-mer counts to their reverse
      complement and vice versa.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Smoothing.}

  How do we compare multiple $k$-mer profiles?
  \medskip

  One sample might be sequenced deeper than the other, so the optimal $k$ might
  differ between comparisons.
  \pause

  \begin{figure}
    \fbox{
      \hspace{0.5cm}
      \begin{tabular}{ll}
        \bt{GTAAGTAA} & $0$ \\
        \bt{GTAAGTAC} & $1$ \\
        \bt{GTAAGTAG} & $0$ \\
        \bt{GTAAGTAT} & $1$ \\
        \bt{GTAAGTCA} & $9$ \\
        \bt{GTAAGTCC} & $8$ \\
        \bt{GTAAGTCG} & $3$ \\
        \bt{GTAAGTCT} & $7$ \\
      \end{tabular}
      \hspace{0.5cm}
      $\Rightarrow$
      \hspace{0.5cm}
      \begin{tabular}{ll}
        \bt{GTAAGTA}  & $2$ \\
        \bt{GTAAGTCA} & $9$ \\
        \bt{GTAAGTCC} & $8$ \\
        \bt{GTAAGTCG} & $3$ \\
        \bt{GTAAGTCT} & $7$ \\
      \end{tabular}
      \hspace{0.5cm}
    }
    \caption{Smoothing by collapsing sub-profiles.}
  \end{figure}
\end{frame}

\begin{frame}
  \frametitle{Smoothing.}

  The function to determine when to smooth is a parameter:

  \begin{itemize}
    \item Median.
    \item Minimum.
    \item Average.
    \item \ldots
  \end{itemize}
  \bigskip

  This function has a threshold, which is also a parameter.
  \bigskip
  \pause

  We can index with a large $k$-mer size, this method automatically uses the
  optimal size when comparing.
\end{frame}

\section{kLib}
\begin{frame}
  \frametitle{Library for $k$-mer profiles.}

  \bt{kLib}:
  \begin{itemize}
    \item Analyse a fasta file.
    \item Loading and saving of profiles.
    \item Merging.
    \item Balancing / splitting.
  \end{itemize}
  \bigskip
  \pause

  \bt{kDiffLib}:
  \begin{itemize}
    \item Smoothing.
    \item Make a comparison recipe.
  \end{itemize}
  \bigskip
  \pause

  \bt{kMer}:
  \begin{itemize}
    \item Command line interface.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{A more general library.}

  These functions are not limited to $k$-mer profiles.
  \bigskip

  \bt{metrics}:
  \begin{itemize}
    \item Scaling.
    \item Extracting positive counts.
    \item Multiset distance function.
    \begin{itemize}
      \item Pairwise distance functions.
    \end{itemize}
    \item Euclidean distance function.
    \item Summary functions for the smoothing algorithm.
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{API documentation.}
\end{frame}

\section{Questions?}
\lastpagetemplate
\begin{fframe}
  \begin{center}
    Acknowledgements:
    \bigskip
    \bigskip

    Yahya Anvar

  \end{center}

  \vfill
  \bs{https://humgenprojects.lumc.nl/svn/k-mer/}
\end{fframe}

\end{document}
